src/v/kafka/server/connection_context.cc

```diff
@@ -353,169 +356,137 @@ connection_context::reserve_request_units(api_key key, size_t size) {
+        // snc_quota_mgr().get_or_create_quota_context() in
+        // record_tp_and_calculate_throttle(), but there is possibility
+        // that the changing configuration could still take us into this
+        // branch with unmatching (and even null) _snc_quota_context.
+        // Simply an unmatching _snc_quota_context is no big deal because
+        // it is a one off event, but we need protection from it being
+        // nullptr
+        if (likely(_snc_quota_context)) {
+            _server.snc_quota_mgr().record_request_intake(
+              *_snc_quota_context, size);
+        }
+    }
+
+    auto sres = ss::make_lw_shared(std::move(sres_in));
+
+    auto remaining = size - request_header_size - hdr.client_id_buffer.size()
+                     - hdr.tags_size_bytes;
+    auto buf = co_await read_iobuf_exactly(conn->input(), remaining);
+    if (_server.abort_requested()) {
+        // _server._cntrl etc might not be alive
+        co_return;
+    }
+    auto self = shared_from_this();
+    auto rctx = request_context(
+      self, std::move(hdr), std::move(buf), sres->backpressure_delay);
+    /*
+     * we process requests in order since all subsequent requests
+     * are dependent on authentication having completed.
+     *
+     * the other important reason for disabling pipeling is because
+     * when a sasl handshake with version=0 is processed, the next
+     * data on the wire is _not_ another request: it is a
+     * size-prefixed authentication payload without a request
+     * envelope, and requires special handling.
+     *
+     * a well behaved client should implicitly provide a data stream
+     * that invokes this behavior in the server: that is, it won't
+     * send auth data (or any other requests) until handshake or the
+     * full auth-process completes, etc... but representing these
+     * nuances of the protocol _explicitly_ in the server makes its
+     * behavior easier to understand and avoids misbehaving clients
+     * creating server-side errors that will appear as a corrupted
+     * stream at best and at worst some odd behavior.
+     */
+
+    const auto correlation = rctx.header().correlation;
+    const sequence_id seq = _seq_idx;
+    _seq_idx = _seq_idx + sequence_id(1);
+    auto res = kafka::process_request(
+      std::move(rctx), _server.smp_group(), *sres);
+
+    /*
+     * first stage processed in a foreground.
+     *
+     * if the dispatch/first stage failed, then we need to
+     * need to consume the second stage since it might be
+     * an exceptional future.
+     */
+    auto dispatch = co_await ss::coroutine::as_future(
+      std::move(res.dispatched));
+    if (dispatch.failed()) {
+        try {
+            co_await std::move(res.response);
+        } catch (...) {
+            vlog(
+              klog.info,
+              "Discarding second stage failure {}",
+              std::current_exception());
+        }
+        vlog(
```

Comment:
question: should we move this log before we wait the response? It seems like it would make sense to see the second stage failure after the error dispatch...

Reply:
Nice spot. I wonder if anyone ever noticed this in the logs. Presumably it would be showing up as an error in dispatch then an error in the first stage. Hmm. Anyway, I fixed based on your suggestion.